Adding plugin support

jtaubensee opened the pull request:
## Description
Addresses #106.

Adds initial plugin points for all clients / sender receivers. This will be updated once the design in finalized.

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] **I have read the [contribution guidelines](./CONTRIBUTING.md).**
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR.
- [x] The pull request does not introduce breaking changes (unless a major version change occurs in the assembly and module).
- [x] If applicable, the public code is properly documented.
- [x] Pull request includes test coverage for the included changes.
- [x] The code builds without any errors.

jtaubensee commented:
This is still very much a work in progress, but wanted to keep the ball rolling.

Here are some things that still need thought:
* Exception handling - I was thinking we could do something like [`OnMessageOptions.ExceptionReceived`](https://docs.microsoft.com/dotnet/api/microsoft.servicebus.messaging.onmessageoptions.exceptionreceived?view=azureservicebus-4.0.0)
  * Should we allow you to implement a custom exception that would allow us to send the message anyway? For example, if we were using KeyVault, we wouldn't want to send an unencrypted message, but if we were just doing something with performance counters, we'd still like to send, even if our custom logic failed.
* What other extension points are needed? Right now, I have:
  * BeforeMessageSend
  * AfterMessageReceive
* When can the plugins be registered? https://github.com/Azure/azure-service-bus-dotnet/issues/106#issuecomment-298720362

SeanFeldman commented:
> Should we allow you to implement a custom exception that would allow us to send the message anyway?

Custom exception or exception controlling callback that allows to prevent exception from bubbling up?

> When can the plugins be registered?

I would advocate at configuration time only.

> What other extension points are needed?

If you treat it as a pipeline, perhaps should have `Before` and 'After`? For the outgoing  pipeline you could have before and after, that would be used to stamp the brokered message before it hits the wire.

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs:
```diff
@@ -211,24 +214,44 @@ public Task<IList<Message>> ReceiveAsync(int maxMessageCount)
+                        var processedMessage = message;
+                        foreach (var afterEachMessageReceiveTask in this.AfterEachMessageReceiveTasks)
+                        {
+                            processedMessage = await afterEachMessageReceiveTask(message);
```
configureawait(false)

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs:
```diff
@@ -110,10 +113,30 @@ public Task SendAsync(Message message)
+                        var processedMessage = message;
+                        foreach (var beforeEachMessageSendTask in this.BeforeEachMessageSendTasks)
+                        {
+                            processedMessage = await beforeEachMessageSendTask(message);
```
configureawait(false)

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs:
```diff
@@ -148,10 +171,19 @@ public Task SendAsync(Message message)
+                {
+                    foreach (var beforeEachMessageSendTask in this.BeforeEachMessageSendTasks)
+                    {
+                        processedMessage = await beforeEachMessageSendTask(message);
```
configureawait(false)

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs:
```diff
@@ -102,6 +103,8 @@ protected MessageReceiver(ReceiveMode receiveMode, TimeSpan operationTimeout, Re
             this.messageReceivePumpSyncLock = new object();
         }
 
+        public IList<Func<Message, Task<Message>>> AfterEachMessageReceiveTasks { get; set; }
```
public setter?

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs:
```diff
@@ -71,8 +71,11 @@ public class MessageSender : ClientEntity, IMessageSender
+            this.BeforeEachMessageSendTasks = new List<Func<Message, Task<Message>>>();
         }
 
+        public IList<Func<Message, Task<Message>>> BeforeEachMessageSendTasks { get; set; }
```
public setter?

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs:
```diff
@@ -211,24 +214,44 @@ public Task<IList<Message>> ReceiveAsync(int maxMessageCount)
                     }, serverWaitTime)
                     .ConfigureAwait(false);
+
+                if (this.AfterEachMessageReceiveTasks == null)
```
How about adding a default noop task and skip the evaluation?

jtaubensee replied:
@SeanFeldman - I think a no-op makes sense here, still working on the best way to do it though.

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs:
```diff
@@ -110,10 +113,30 @@ public Task SendAsync(Message message)
             {
+                IList<Message> processedMessageList;
+
+                if (this.BeforeEachMessageSendTasks == null)
```
same idea with noop

jtaubensee commented:
Just added an ExceptionHandler for `ServiceBusPlugin`. This probably still needs a few things:
* Do we want to have a custom exception, or just reuse the `ServiceBusException.IsTransient` logic? I'm guessing we want something else
* Do we want to reuse `ExceptionReceivedEventArgs` from `MessageHandlerOptions`?

jtaubensee commented:
Regarding exception handling, any exception in a plugin will be thrown to the corresponding action. So you will get an exception on `QueueClient.SendAsync`, unless you throw a `ServiceBusPluginException` with `ShouldCompleteOperation` set to true (in your plugin code). The idea here is that, you wouldnt want to send an unencrypted message if you can't reach KeyVault, but you might be ok with the message not being compressed.

Outstanding items to be addressed:
* Only registering plugins at startup. @SeanFeldman - Were you thinking this would only be accessible through the constructor then?
* Additional extension points

@nemakam / @vinaysurya please take a look as well.

SeanFeldman commented:
@jtaubensee registration - the thought was to discourage registration of plugins during runtime. With no clear configuration and execution phases, don't see how it would be possible other than during construction.

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs:
```diff
@@ -960,5 +1009,48 @@ void ThrowIfNotPeekLockMode()
+            this.RegisteredPlugins.Add(serviceBusPlugin);
+        }
+
+        public void UnregisterPlugin(ServiceBusPlugin serviceBusPlugin)
```
What would be the case to unregister a plugin? Curious to understand.

jtaubensee replied:
This might be a stretch, and maybe that means its not needed for now, but I can think of a scenario where I would want to unregister when I continually receive errors with my plugin, and I want to continue using the same client...

SeanFeldman replied:
> This might be a stretch, and maybe that means its not needed for now

Then don't add it 😃 

>  I can think of a scenario where I would want to unregister when I continually receive errors with my plugin

I'm not sure about this. If you have a component that you rely on and it misbehaves, you would find the issue and redeploy the component. I might be off on this one, but feels like adding this w/o actual need would be premature.

jtaubensee replied:
I can get behind that. Any thoughts @nemakem / @vinaysurya?

nemakam replied:
I partly agree with @Sean here. I think we should be able to register plugin only during creation of Queue unless we see a strong reason to change things dynamically.
But at the same time, we might want to have `UnregisterPlugin` or `ClosePlugin` equivalent so that plugin can probably release some resources of its own if required. But this would require us to expose additional methods on plugin. 
So in effect, we might end up having two more functions `OnPluginRegistered` and `OnPluginUnregistered` methods in `ServiceBusPlugin`

---
In reply to: [117356385](https://github.com/Azure/azure-service-bus-dotnet/pull/153#discussion_r117356385) [](ancestors = 117356385)

nemakam replied:
But now to think of it, i can think of scenarios where we may want to unregister plugin. 
Lets say we have a plugin to tackle inter-op scenarios for users to switch from old client to new client. At some point in time they realise that they are done with the interop handling (upgrade scenarios), and now want to unregister the plugin. 

---
In reply to: [117356385](https://github.com/Azure/azure-service-bus-dotnet/pull/153#discussion_r117356385) [](ancestors = 117356385)

SeanFeldman replied:
Interesting, saw a comment from @nemakam, but now it's gone.
Reason I'm against this method for now is YAGNI. If at some point in time a plugin is no longer needed, it should be removed, not just disabled.

nemakam replied:
Yeah, I cant find my own comment. Reposting it here:

>I partly agree with @Sean here. I think we should be able to register plugin only during creation of Queue unless we see a strong reason to change things dynamically. But at the same time, we might want to have UnregisterPlugin or ClosePlugin equivalent so that plugin can probably release some resources of its own if required. But this would require us to expose additional methods on plugin. So in effect, we might end up having two more functions OnPluginRegistered and OnPluginUnregistered methods in ServiceBusPlugin

>But now to think of it, i can think of scenarios where we may want to unregister plugin. Lets say we have a plugin to tackle inter-op scenarios for users to switch from old client to new client. At some point in time they realise that they are done with the interop handling (upgrade scenarios), and now want to unregister the plugin.

nemakam replied:
And regarding YAGNI, I think we might end up working on building an interop plugin very soon. Given that, I feel we could leave this UnRegister as-is and not remove it as we don't see any issues with it other than YAGNI.

>  If at some point in time a plugin is no longer needed, it should be removed, not just disabled.
Based on current implementation, IMO  `UnregisterPlugin` seems to be "removing plugin" and not just "disabling" it.

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs:
```diff
@@ -90,6 +92,47 @@ public class MessageSender : ClientEntity, IMessageSender
             }
         }
 
+        private async Task<Message> ProcessMessages(Message message)
```
Method name `ProcessMessages`, but it only processes a single message.

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs:
```diff
@@ -90,6 +92,47 @@ public class MessageSender : ClientEntity, IMessageSender
+                {
+                    processedMessage = await plugin.BeforeMessageSend(message).ConfigureAwait(false);
+                }
+                catch (Exception e)
```
Same exception filtering as above could be used.

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/QueueClient.cs:
```diff
@@ -305,5 +305,23 @@ public Task CancelScheduledMessageAsync(long sequenceNumber)
+            this.InnerReceiver.UnregisterPlugin(serviceBusPlugin);
+        }
+
+        public void UnregisterPlugin(Type serviceBusPluginType)
```
Is this overload needed?

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/SubscriptionClient.cs:
```diff
@@ -308,5 +308,20 @@ public Task AddRuleAsync(string ruleName, Filter filter)
+            this.InnerSubscriptionClient.InnerReceiver.UnregisterPlugin(serviceBusPlugin);
+        }
+
+        public void UnregisterPlugin(Type serviceBusPluginType)
```
Same question, is this overload needed?

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/TopicClient.cs:
```diff
@@ -142,5 +142,20 @@ public Task CancelScheduledMessageAsync(long sequenceNumber)
+            this.InnerSender.UnregisterPlugin(serviceBusPlugin);
+        }
+
+        public void UnregisterPlugin(Type serviceBusPluginType)
```
same as above

nemakam commented on src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs:
```diff
@@ -160,6 +162,40 @@ internal set
+        }
+
+        private async Task<IList<Message>> ProcessMessages(IList<Message> messageList)
+        {
```
return right here if there are no registered plugins. need to iterate through all the messages.

nemakam commented on src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs:
```diff
@@ -160,6 +162,40 @@ internal set
+            {
+                try
+                {
+                    processedMessage = await plugin.AfterMessageReceive(message).ConfigureAwait(false);
```
>processedMessage = await plugin.AfterMessageReceive(message).ConfigureAwait(false); [](start = 20, length = 83)

add log statements before and after calling plugin

nemakam commented on src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs:
```diff
@@ -160,6 +162,40 @@ internal set
+                }
+                catch (Exception ex) when (ex is ServiceBusPluginException pluginException && pluginException.ShouldCompleteOperation)
+                {
+                    continue;
```
>continue; [](start = 20, length = 9)

should log here as well

nemakam commented on src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs:
```diff
@@ -90,6 +92,40 @@ public class MessageSender : ClientEntity, IMessageSender
+            {
+                try
+                {
+                    processedMessage = await plugin.BeforeMessageSend(message).ConfigureAwait(false);
```
>processedMessage = await plugin.BeforeMessageSend(message).ConfigureAwait(false); [](start = 20, length = 81)

like in receiver, log before and after calling the plugin to debug.

nemakam commented on src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs:
```diff
@@ -90,6 +92,40 @@ public class MessageSender : ClientEntity, IMessageSender
+            return processedMessage;
+        }
+
+        private async Task<IList<Message>> ProcessMessages(IList<Message> messageList)
```
>private async Task<IList<Message>> ProcessMessages(IList<Message> messageList) [](start = 8, length = 78)

return quick if no plugins found instead of iterating through the messages

nemakam commented on src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs:
```diff
@@ -90,6 +92,40 @@ public class MessageSender : ClientEntity, IMessageSender
+                }
+                catch (Exception ex) when (ex is ServiceBusPluginException pluginException && pluginException.ShouldCompleteOperation)
+                {
+                    continue;
```
>continue; [](start = 20, length = 9)

like in receiver, log the exception

nemakam commented on src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs:
```diff
@@ -90,6 +92,40 @@ public class MessageSender : ClientEntity, IMessageSender
+                {
+                    processedMessage = await plugin.BeforeMessageSend(message).ConfigureAwait(false);
+                }
+                catch (Exception ex) when (ex is ServiceBusPluginException pluginException && pluginException.ShouldCompleteOperation)
```
>pluginException.ShouldCompleteOperation [](start = 94, length = 39)

What are your opinions on exposing this on the plugin? Who should decide whether this is an exception that impacts or not? - User or the plugin developer?

SeanFeldman replied:
My gut feeling tells me the user. Plugin is there to do the job, not to decide on the overall course of actions.

nemakam commented on src/Microsoft.Azure.ServiceBus/Primitives/ServiceBusPluginException.cs:
```diff
@@ -0,0 +1,41 @@
+    /// </summary>
+    public class ServiceBusPluginException : Exception
+    {
+        public ServiceBusPluginException(bool shouldCompleteOperation)
```
>shouldCompleteOperation [](start = 46, length = 23)

maybe rename to `shouldContinueOperation`?

SeanFeldman commented:
Agree that UnregisterPlugin would achieve the same result, but why would
one register a plugin to unregister it shortly after if plugin is not
needed :D​

jtaubensee commented:
@SeanFeldman / @nemakam - Ok, I think I've incorporated most of your feedback now. Here are some points to clarify:
* I've removed the `ServiceBusPluginException`, and added a bool to the `ServiceBusPlugin`
  * `public virtual bool ShouldContinueOnException => false;` - this will handle whether or not the error should continue or throw
* I've also added an abstract name that is required. This allows us to identify plugins easier;
  * `public abstract string Name { get; }`
* Added logging

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs:
```diff
@@ -160,6 +162,51 @@ internal set
+            {
+                try
+                {
+                    MessagingEventSource.Log.PluginCalled(plugin.Name, message.MessageId);
```
`PluginCallStarted` & `PluginCallCompleted`?

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs:
```diff
@@ -160,6 +162,51 @@ internal set
+                }
+                catch (Exception ex)
+                {
+                    MessagingEventSource.Log.PluginException(plugin.Name, message.MessageId, ex.Message);
```
`PluginCallFailed` or `PluginCallThrewException`?

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs:
```diff
@@ -960,5 +1011,39 @@ void ThrowIfNotPeekLockMode()
+            }
+            if (this.RegisteredPlugins.Any(p => p.Name == serviceBusPluginName))
+            {
+                var plugin = this.RegisteredPlugins.First(p => p.Name == serviceBusPluginName);
```
The way I interpret this is that you can have more than one plugin of different types with the same name. That doesn't sound right.

jtaubensee replied:
Good catch - it should be this: `this.RegisteredPlugins.Any(p => p.Name == serviceBusPlugin.Name)`

The name should be unique, so if you start using our KeyVault plugin the name/id will be set by the plugin itself, so you won't be able to register it multiple times.

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs:
```diff
@@ -960,5 +1011,39 @@ void ThrowIfNotPeekLockMode()
             MessagingEventSource.Log.RegisterOnMessageHandlerStop(this.ClientId);
         }
+
+        public void RegisterPlugin(ServiceBusPlugin serviceBusPlugin)
```
What about name that is already used?

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Core/ServiceBusPlugin.cs:
```diff
@@ -0,0 +1,24 @@
+
+    public abstract class ServiceBusPlugin
+    {
+        public abstract string Name { get; }
```
Should it default to something like a new Guid?

jtaubensee replied:
I would expect that this be set by the plugin, so we wouldnt want to auto generate this. We are hoping to prevent users from calling use multiple times with the same plugin.

jtaubensee commented:
@SeanFeldman - 
> I don't see plugin work been done differently based on the fact if it's a sender or a receiver

They are largely the same however, the one difference is the extension point that is called.
* Sender: `plugin.BeforeMessageSend`
* Receiver: `plugin.AfterMessageReceive`

SeanFeldman commented:
Right @jtaubensee, the invocation takes place at different places, but the registration logic remains the same. That's why I felt the configuration part could be the same.

SeanFeldman commented:
PS: i can raise a PR tomorrow. Would that work?

nemakam commented on src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs:
```diff
@@ -201,6 +199,46 @@ internal set
+
+        private async Task<IList<Message>> ProcessMessages(IList<Message> messageList)
+        {
+            if (!this.RegisteredPlugins.Any())
```
>.Any() [](start = 39, length = 6)

Use RegisteredPlugins.Count instead of Any() since this is in the hot-path and Any() is considerably slower than Count

danielmarbach commented on src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs:
```diff
@@ -160,6 +162,51 @@ internal set
+
+        private async Task<IList<Message>> ProcessMessages(IList<Message> messageList)
+        {
+            if (this.RegisteredPlugins == null || !this.RegisteredPlugins.Any())
```
Can this ever be null? Wouldn't be more efficient to make sure we always have a list there and use the count property instead of linq?

SeanFeldman replied:
[Can't be null](https://github.com/Azure/azure-service-bus-dotnet/blob/6cc56bda017d28737b1eae47124cd6d115bc2086/src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs#L132).

vinaysurya replied:
@danielmarbach, looks like this is already corrected in the checked code
https://github.com/Azure/azure-service-bus-dotnet/blob/6cc56bda017d28737b1eae47124cd6d115bc2086/src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs#L227

danielmarbach commented on src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs:
```diff
@@ -90,6 +92,51 @@ public class MessageSender : ClientEntity, IMessageSender
+
+        private async Task<IList<Message>> ProcessMessages(IList<Message> messageList)
+        {
+            if (this.RegisteredPlugins == null || !this.RegisteredPlugins.Any())
```
Same question as above

SeanFeldman replied:
Same answer with 👍 for `IList<T>.Count`

vinaysurya replied:
Same here, looks already taken care of in current code
